### ISSUE TITLE
log.txt not getting saved in COLAB

### ISSUE THREAD

**Issue by Rahul-Venugopal** [AlexeyAB/darknet#2398]
Hi ,

I am using following command to train yolov3 in google COLAB 
`! ./darknet detector train data/obj.data cfg/yolov3-obj_r1.cfg backup/yolov3-obj_r1_8000.weights  -dont_show -map | tee log.txt`

Since the COLAB connection gets disconnected after sometime , I will not be able to complete full iteration at one time . When I check `chart.png` , it gets saved and updated in **every 100 iterations** but unfortunately `log.txt `is not getting updated after 100 iterations and also it remains as an empty file if the training stops in between.  

It is noted that log.txt get saved **only** if I complete full iterations and else it remains as an empty file. 
 
**Is it possible to save and update log.txt as same way as that of Chart.png ?**

Thanks 
Rahul

**Comment by AlexeyAB** [comment]
@Rahul-Venugopal Hi,

Try to use `unbuffer -p tee log.txt`. 

So install it if it is missing
```
sudo apt-get install unbuffer
```

And use such command:
`./darknet detector train data/obj.data cfg/yolov3-obj_r1.cfg backup/yolov3-obj_r1_8000.weights -dont_show -map | unbuffer -p tee log.txt`

**Comment by Rahul-Venugopal** [comment]
Hi @AlexeyAB  ,

Thanks for your reply and it helped 👍  

If I have train.txt and valid.txt with seperate set of images , **which avg.loss** gets plotted in the **chart.png** ?
For an iteration the result is , 
1802: `0.513317, 0.332773 avg loss`, 0.001000 rate, 13.871005 seconds, 115328 images

are this `(loss on training set , loss on validation set )`  ? 



( PS :` !sudo apt install expect` helped me to install unbuffer in COLAB. If it helps anyone in the future )

**Comment by AlexeyAB** [comment]
@Rahul-Venugopal Hi,

* Loss is always for Training dataset `train=` in obj.data file
* Accuracy (mAP/topk/...) is always for Validation dataset `valid=` in obj.data file

> In addition to that , is it possible to get ` Last accuracy mAP@0.5` on both training set and validation set separately in the log file ?

No.
If you want to get accuracy for Training dataset, then you should set `valid=train.txt` in obj.data file.

**Comment by Rahul-Venugopal** [comment]
Hi @AlexeyAB  ,

Thanks for clearing the doubt . 

> * Loss is always for Training dataset `train=` in obj.data file

if so , can you please tell me what does two values **0.085398** and  **0.113790** in the following output means ? (As I understood 0.113790 is avg loss on training set , then what does 0.085398 represent ?) 

Last accuracy mAP@0.5 = 80.85 % 
 `10357: 0.085398, 0.113790 avg loss`, 0.000010 rate, 6.457871 seconds, 662848 images
Loaded: 0.000075 seconds


> * Accuracy (mAP/topk/...) is always for Validation dataset `valid=` in obj.data file

Actually I was trying to plot like the following figure 

![cnn_keras_plot](https://user-images.githubusercontent.com/30955455/52805609-3cc87880-3087-11e9-989c-5f2b3efa1eb4.png)

As I understood , I **cannot** plot a graph like this with YOLOv3. Please correct me if I am wrong.

**Comment by AlexeyAB** [comment]
@Rahul-Venugopal 

> (As I understood 0.113790 is avg loss on training set , then what does 0.085398 represent ?)
0.113790 is average loss
0.085398 is loss for the current batch


> As I understood , I cannot plot a graph like this with YOLOv3. 

Yes, you can't.
